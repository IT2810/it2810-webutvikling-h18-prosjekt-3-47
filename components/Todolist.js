import React, { Component } from "react";
import {
  AppRegistry,
  StyleSheet,
  Text,
  View,
  FlatList,
  AsyncStorage,
  Button,
  TextInput,
  Keyboard,
  Platform
} from "react-native";

//Sjekker hvilken plattform vi er på
const isAndroid = Platform.OS === "android";
const viewPadding = 10;

export default class Todolist extends Component {
  //Lager default state
  state = {
    tasks: [],
    text: ""
  };
  
  //Funksjon som endrer tilstanden til text
  changeTextHandler = text => {
    this.setState({ text: text });
  };
  //Funksjon som legger til en ny oppgave og lagrer dette ved hjelp av funksjonen save
  addTask = () => {
    //Sjekker om det er noe tekst og setter den boolske verdien i notEmpty
    let notEmpty = this.state.text.trim().length > 0;

    //Hvis notEmpty er true kan vi lage en ny task
    if (notEmpty) {
      this.setState(
        prevState => {
          let { tasks, text } = prevState;
          return {
            tasks: tasks.concat({ key: Date.now().toString() + '_' + tasks.length.toString(), text: text }),

            text: ""
          };
        },
        () => Tasks.save(this.state.tasks)
      );
    } 
  };
  //Denne funksjonen sletter en oppgave og lagrer resultatet av tilstanden
  deleteTask = i => {
    if(Number.isInteger(Number(i))){
      this.setState(
        prevState => {
          let tasks = prevState.tasks.slice();

          tasks.splice(i, 1);

          return { tasks: tasks };
        },
        () => Tasks.save(this.state.tasks)
      );
    }
  };

  componentDidMount() {
    //Viser keyboard ut ifra plattform
    Keyboard.addListener(
      isAndroid ? "keyboardDidShow" : "keyboardWillShow",
      e => this.setState({ viewPadding: e.endCoordinates.height + viewPadding })
    );
    //Gjemmer keyboard ut ifra plattform
    Keyboard.addListener(
      isAndroid ? "keyboardDidHide" : "keyboardWillHide",
      () => this.setState({ viewPadding: viewPadding })
    );

    Tasks.all(tasks => this.setState({ tasks: tasks || [] }));
  }

  render() {
    return (
      <View
      //Henter style fra stylesheet lenger ned
        style={[styles.container, { paddingBottom: this.state.viewPadding }]}
      >
        <FlatList
          style={styles.list}
          data={this.state.tasks}
          renderItem={({ item, index }) =>
            <View>
              <View style={styles.listItemCont}>
                <Text style={styles.listItem}>
                  {item.text}
                </Text>
                <Button style={styles.button} title="X" onPress={() => this.deleteTask(index)} />
              </View>
              <View style={styles.hr} />
            </View>}
        />
        <TextInput
          style={styles.textInput}
          onChangeText={this.changeTextHandler}
          onSubmitEditing={this.addTask}
          value={this.state.text}
          placeholder="Add Tasks"
          returnKeyType="done"
          returnKeyLabel="done"
        />
      </View>
    );
  }
}

let Tasks = {
  // Denne funksjone lager en liste med objekter
  convertToArrayOfObject(tasks, callback) {
    //console.log('convertToArrayOfObject');
    //console.log(tasks);
    //console.log(tasks ? tasks.split("||").map((task, i) => ({ key: i.toString(), text: task })) : [])
    return callback(
      tasks ? tasks.split("||").map((task, i) => ({ key: Date.now().toString() + '_' + i.toString() , text: task })) : []
    );
  },
  //Lager en streng med teksten til objektene
  convertToStringWithSeparators(tasks) {
    //console.log('convertToStringWithSeparators');
    return tasks.map(task => task.text).join("||");
  },
  //Henter oppgaver ved hjelp av AsyncStorage
  all(callback) {
    //console.log('all');
    //console.log('callback:', callback);
    return AsyncStorage.getItem("TASKS", (err, tasks) =>
      this.convertToArrayOfObject(tasks, callback)
    );
  },
  //Lagrer oppgaver ved hjelp av AsyncStorage
  save(tasks) {
    //console.log('save');
    AsyncStorage.setItem("TASKS", this.convertToStringWithSeparators(tasks));
  }
};



//For å style elementene til todolist
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#F5FCFF",
    padding: viewPadding,
    paddingTop: 20
  },
  list: {
    width: "100%"
  },
  listItem: {
    paddingTop: 2,
    paddingBottom: 2,
    fontSize: 18,
    width: "90%"
  },
  hr: {
    height: 1,
    backgroundColor: "gray"
  },
  listItemCont: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between"
  },
  textInput: {
    height: 40,
    paddingRight: 10,
    paddingLeft: 10,
    borderColor: "gray",
    borderWidth: isAndroid ? 0 : 1,
    width: "100%"
  }
});

AppRegistry.registerComponent("TodoList", () => TodoList);

module.exports = Todolist;
