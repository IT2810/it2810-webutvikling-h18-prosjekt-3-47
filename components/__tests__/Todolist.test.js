import {View} from 'react-native';
import React from 'react';
import Todolist from '../Todolist';
import renderer from 'react-test-renderer';



describe('TodoList Component', () => {

    /*
    it('should render new tasks when they are added', () => {
        const wrapper = shallow(<View />);

        const firstView = wrapper.find('view').at(0);
        const secondView = wrapper.find('view').at(1);

        expect(firstView.props().style.toEqual([styles.container, { paddingBottom: this.state.viewPadding }]));

    });

    */

    it('renders correctly', () => {
        const todos = renderer.create(<Todolist />);

        expect(todos.toJSON()).toMatchSnapshot();
    });


    it('should have empty tasks in state when first loading', () => {
        const todoLists = renderer.create(<Todolist />).getInstance();

        expect(todoLists.state.tasks).toEqual([])
    });

    it('should have empty text in state when first loading', () => {
        const todoLists = renderer.create(<Todolist />).getInstance();

        expect(todoLists.state.text).toEqual('')
    });

    it('should not add a new task when addTask is called and text is empty', () => {
        const todoLists = renderer.create(<Todolist />).getInstance();

        // state.text is at default empty string value, so no tasks should be created by addTask
        todoLists.addTask();
        expect(todoLists.state.tasks).toEqual([]);


        // Adds a lot of whitespace to state.text, then runs addTask
        todoLists.setState({
            text: '     ' +
                '   '
        }, () => {
            todoLists.addTask();
        });

        // addTask should filter on whitespace, so no tasks should be created.
        expect(todoLists.state.tasks).toEqual([]);
    });


    it('should add a new task when addTask is called and there is text', () => {
        const todoLists = renderer.create(<Todolist />);
        const todoListsInstance = todoLists.getInstance();
        const todoListInstanceRendered = renderer.create(todoListsInstance.instance);

        todoListsInstance.setState({
            text: 'TODO-dkfkdfjfjjdf-1'
        }, () => {
            todoListsInstance.addTask();
        });

        expect(todoListsInstance.state.tasks).toEqual([
            {
                key: '0',
                text: 'TODO-dkfkdfjfjjdf-1'
            }
        ]);

        // expect(todoLists.findByType())

        console.log(todoLists.toJSON());
        console.log(todoLists.toJSON().children);
        //console.log(todoListsInstance);


        /*

        console.log(renderer.create(todoListsInstance.instance));
        console.log(todoListInstanceRendered);
        console.log(todoListInstanceRendered.toJSON());
        console.log(todoListInstanceRendered.toTree());

        */


        expect(todoListInstanceRendered.toJSON()).toMatchSnapshot();
    });

    it('should clear state.text when adding a task', () => {
        const todoLists = renderer.create(<Todolist />).getInstance();

        todoLists.setState({
            text: 'TODO-jdjdhfhhhfncnc-2'
        }, () => {
            todoLists.addTask();
        });

        // Expecting a task to be added
        expect(todoLists.state.tasks).toEqual([
            {
                key: '0',
                text: 'TODO-jdjdhfhhhfncnc-2'
            }
        ]);

        // Expecting state.text to be cleared by addTask
        expect(todoLists.state.text).toEqual('');
    });

    it('should add several tasks', () => {
        const todoLists = renderer.create(<Todolist />).getInstance();

        todoLists.setState({
            text: 'TODO-ssdfodu33sdjsdh-1'
        }, () => {
            todoLists.addTask();
        });

        // Expecting there to be one task
        expect(todoLists.state.tasks).toEqual([
            {
                key: '0',
                text: 'TODO-ssdfodu33sdjsdh-1'
            }
        ]);

        todoLists.setState({
            text: 'TODO-ssdfodu33sdjsdh-2'
        }, () => {
            todoLists.addTask();
        });

        // Expecting there to be two tasks
        expect(todoLists.state.tasks).toEqual([
            {
                key: '0',
                text: 'TODO-ssdfodu33sdjsdh-1'
            },
            {
                key: '1',
                text: 'TODO-ssdfodu33sdjsdh-2'
            }
        ]);

        todoLists.setState({
            text: 'TODO-ssdfodu33sdjsdh-3'
        }, () => {
            todoLists.addTask();
        });

        // Expecting there to be three tasks
        expect(todoLists.state.tasks).toEqual([
            {
                key: '0',
                text: 'TODO-ssdfodu33sdjsdh-1'
            },
            {
                key: '1',
                text: 'TODO-ssdfodu33sdjsdh-2'
            },
            {
                key: '2',
                text: 'TODO-ssdfodu33sdjsdh-3'
            }
        ]);
    });

});


