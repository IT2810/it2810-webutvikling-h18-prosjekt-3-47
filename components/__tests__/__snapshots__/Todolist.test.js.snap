// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "justifyContent": "center",
        "padding": 10,
        "paddingTop": 20,
      },
      Object {
        "paddingBottom": undefined,
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Add Tasks"
    returnKeyLabel="done"
    returnKeyType="done"
    style={
      Object {
        "borderColor": "gray",
        "borderWidth": 1,
        "height": 40,
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": "100%",
      }
    }
    value=""
  />
</View>
`;

exports[`TodoList Component should add a new task when addTask is called and there is text 1`] = `null`;
